lib/cib/cib_utils.c

```diff
@@ -77,6 +78,259 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+
+                return true;
+            }
+
+        } else {
+            // Change to an existing object
+            if (regex) {
+                if (pcmk__str_eq(diff_xpath, obj_xpath, pcmk__str_regex)) {
+                    return true;
+                }
+
+            } else if (pcmk__starts_with(diff_xpath, obj_xpath)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+/*!
+ * \internal
+ * \brief Check whether a given CIB element was modified in a CIB update (v2)
+ *
+ * Search a CIB diff notification message's XML v2 patchset for the given
+ * element.
+ *
+ * \param[in] patchset  XML patchset from CIB update
+ * \param[in] element   XML tag of CIB element to check (\c NULL is equivalent
+ *                      to \c XML_TAG_CIB). Supported values include any CIB
+ *                      element supported by \c pcmk__cib_abs_xpath_for().
+ *
+ * \return \c true if \p element was modified in the CIB update, or \c false
+ *         otherwise
+ */
+static bool
+element_in_patchset_v2(const xmlNode *patchset, const char *element)
+{
+    const char *element_xpath = pcmk__cib_abs_xpath_for(element);
+    const char *parent_xpath = pcmk_cib_parent_name_for(element);
+
+    CRM_CHECK(element_xpath != NULL, return false); // Unsupported element
+
+    return object_in_patchset_v2(patchset, element, element_xpath, parent_xpath,
+                                 false);
+}
+
+/*!
+ * \internal
+ * \brief Check whether a given CIB element was modified in a CIB update
+ *
+ * Search a CIB diff notification message's XML patchset for the given element.
+ *
+ * \param[in] msg      CIB update result (from a diff notification)
+ * \param[in] element  XML tag of CIB element to check (\c NULL is equivalent to
+ *                     \c XML_TAG_CIB). Supported values include any CIB element
+ *                     supported by \c pcmk__cib_abs_xpath_for().
+ *
+ * \return \c true if \p element was modified in the CIB update, or \c false
+ *         otherwise
+ */
+bool
+cib__element_in_patchset(const xmlNode *msg, const char *element)
+{
+    int rc = pcmk_err_generic;
+    int format = 1;
+    xmlNode *patchset = NULL;
+
+    CRM_ASSERT(msg != NULL);
+
+    if ((crm_element_value_int(msg, F_CIB_RC, &rc) != 0) || (rc != pcmk_ok)) {
+        crm_trace("Ignore failed CIB update: %s (%d)", pcmk_strerror(rc), rc);
+        return false;
+    }
+
+    patchset = get_message_xml(msg, F_CIB_UPDATE_RESULT);
+    CRM_CHECK(patchset != NULL, return false);
+
+    crm_element_value_int(patchset, "format", &format);
+    switch (format) {
+        case 1:
+            return element_in_patchset_v1(patchset, element);
+
+        case 2:
+            return element_in_patchset_v2(patchset, element);
+
+        default:
+            crm_warn("Unknown patch format: %d", format);
+            return false;
+    }
+}
+
+#define REGEX_ID            "\\[@" XML_ATTR_ID "='[^']*'\\]"
```

Comment:
Unfortunately for v2 patchsets, we can't use XPath queries. The best we can do is regex matching against the `path` attribute of each `<change>` element. The path will have `.../node_state[@id='1']` or similar.

Having to do string matching makes all of this clunkier than it otherwise could be.